client/backoff/backoff.go

```diff
@@ -22,37 +22,55 @@ import (
 	"time"
 )
 
-// Backoff specifies the parameters of the backoff algorithm.
+// Backoff specifies the parameters of the backoff algorithm. Works correctly
+// if 0 < Min <= Max <= 2^62 (nanosec), and Factor >= 1.
 type Backoff struct {
-	Min    time.Duration
-	Max    time.Duration
-	Factor float64
-	Jitter bool
-	x      int
+	Min    time.Duration // Duration of the first pause.
+	Max    time.Duration // Max duration of a pause.
+	Factor float64       // The factor of duration increase between iterations.
+	Jitter bool          // Add random noise to pauses.
+
+	delta time.Duration // Current pause duration relative to Min, no jitter.
 }
 
-// Duration returns the time to wait on duration x.
-// Every time Duration is called, the returned value will exponentially increase by Factor
-// until Backoff.Max. If Jitter is enabled, will wait an additional random value between
-// 0 and factor^x * min, capped by Backoff.Max.
+// Duration returns the time to wait on current retry iteration.
+// Every time Duration is called, the returned value will exponentially
+// increase by Factor until Backoff.Max. If Jitter is enabled, will wait an
+// additional random value between 0 and Factor^x * Min, capped by Backoff.Max.
 func (b *Backoff) Duration() time.Duration {
-	// min( min * factor ^ x , max)
-	minNanos := float64(b.Min.Nanoseconds())
-	maxNanos := float64(b.Max.Nanoseconds())
-	nanos := minNanos * math.Pow(b.Factor, float64(b.x))
-	nanos = math.Min(nanos, maxNanos)
+	pause := b.Min + b.delta
+
+	newPause := time.Duration(float64(pause) * b.Factor)
+	if newPause > b.Max || newPause < b.Min { // Multiplication could overflow.
+		newPause = b.Max
+	}
+	b.delta = newPause - b.Min
+
 	if b.Jitter {
-		// Generate a number in the range [0, factor^x * b.Min)
-		r := float64(rand.Int63n(int64(nanos)))
-		nanos += r
+		// Add a number in the range [0, pause).
+		pause += time.Duration(rand.Int63n(int64(pause)))
 	}
-	b.x++
-	return time.Duration(nanos) * time.Nanosecond
+	return pause
 }
 
-// Reset sets the internal iteration count back to 0.
+// Reset sets the internal state back to first iteration.
 func (b *Backoff) Reset() {
-	b.x = 0
+	b.delta = 0
+}
+
+// Recover reduces retry pause duration by the given power of Backoff.Factor.
+// This might be useful when calling Retry multiple times, to gradually adapt
+// to the rate of errors: in some use-cases a successfully returned Retry could
+// indicate that a bigger flow of requests can now be handled.
+func (b *Backoff) Recover(factors int) {
```

Comment:
This feels like the TCP congestion algorithm...  :-)

I suggest allowing `Recover` to not take any arguments. That way it can be called once per successful return. 
Each call of `Recover` should reduce the current latency by a linear factor, not an exponential one.  The desired result is to exponentially backoff during errors, and linearly speed up to find the right performance ceiling.

Reply:
Looks like this has to be thought through, let me clean this PR to be only a refactoring. I will create a separate PR after some testing.